enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/roles/Candidate.java

```diff
@@ -38,124 +38,178 @@
-                    Voting.handleVoteRequest( ctx, outcome, req );
-                    break;
-                }
-
-                outcome.addOutgoingMessage( new RaftMessages.Directed( req.from(),
-                        new RaftMessages.Vote.Response( ctx.myself(), outcome.getTerm(), false ) ) );
-                break;
+                outcome.setLeader( ctx.myself() );
+                Appending.appendNewEntry( ctx, outcome, new NewLeaderBarrier() );
+                Leader.sendHeartbeats( ctx, outcome );
+
+                outcome.setLastLogIndexBeforeWeBecameLeader( ctx.entryLog().appendIndex() );
+                outcome.electedLeader();
+                outcome.renewElectionTimeout();
+                outcome.setNextRole( LEADER );
+                log.info( "Moving to LEADER state at term %d (I am %s), voted for by %s",
+                        ctx.term(), ctx.myself(), outcome.getVotesForMe() );
             }
+            return outcome;
+        }
 
-            case ELECTION_TIMEOUT:
+        @Override
+        public Outcome apply( RaftMessages.Vote.Request req ) throws IOException
+        {
+            if ( req.term() > ctx.term() )
             {
-                log.info( "Failed to get elected. Got votes from: %s", ctx.votesForMe() );
-                if ( !Election.start( ctx, outcome, log ) )
-                {
-                    log.info( "Moving to FOLLOWER state after failing to start election" );
-                    outcome.setNextRole( FOLLOWER );
-                }
-                break;
+                outcome.getVotesForMe().clear();
+                outcome.setNextRole( FOLLOWER );
+                log.info( "Moving to FOLLOWER state after receiving vote request from %s at term %d (I am at %d)",
+                        req.from(), req.term(), ctx.term() );
+                Voting.handleVoteRequest( ctx, outcome, req );
+                return outcome;
             }
 
-            case PRUNE_REQUEST:
+            outcome.addOutgoingMessage( new RaftMessages.Directed( req.from(),
+                    new RaftMessages.Vote.Response( ctx.myself(), outcome.getTerm(), false ) ) );
+            return outcome;
+        }
+
+        @Override
+        public Outcome apply( RaftMessages.Timeout.Election election ) throws IOException
+        {
+            log.info( "Failed to get elected. Got votes from: %s", ctx.votesForMe() );
+            if ( !Election.startRealElection( ctx, outcome, log ) )
             {
-                Pruning.handlePruneRequest( outcome, (RaftMessages.PruneRequest) message );
-                break;
+                log.info( "Moving to FOLLOWER state after failing to start election" );
+                outcome.setNextRole( FOLLOWER );
             }
+            return outcome;
+        }
 
-            default:
-                break;
+        @Override
+        public Outcome apply( RaftMessages.PreVote.Request request ) throws IOException
+        {
+            return outcome;
```

Comment:
It's really good that we expose places were we do nothing, so that it doesn't get hidden.